HEN_HOUSE/egs++/geometry/egs_autoenvelope/volcor.h

```diff
@@ -0,0 +1,663 @@
+ * the volume of those regions. The algorithm is described
+ * on main page of the Auto Envelope documentation. */
+VCResults findRegionsWithInscribed(VCOptions *opts, EGS_BaseGeometry *base,
+                                   vector<EGS_BaseGeometry *> inscribed, vector<EGS_AffineTransform *> transforms) {
+
+    EGS_Timer timer;
+    timer.start();
+
+    VCResults results(opts);
+
+    results.uncorrected_volumes =  getUncorrectedVolumes(base);
+
+    EGS_Vector point;
+
+    EGS_I64 n_in_inscribed = 0;
+
+    //use first inscribed object to test if point is in inscribed geom
+    EGS_BaseGeometry *first_inscribed = inscribed[0];
+    EGS_AffineTransform *first_transform = transforms[0];
+
+    HitCounterT hit_counter;
+
+    for (EGS_I64 i=0; i < opts->npoints; i++) {
+
+        point = opts->getRandomPoint();
+
+        EGS_Vector inscribed_point(point);
+        first_transform->transform(inscribed_point);
+
+        if (!first_inscribed->isInside(inscribed_point)) {
+            continue;
+        }
+        n_in_inscribed += 1;
+
+        // transform to every location and check which regions in the base it is
+        for (size_t sidx = 0; sidx < transforms.size();  sidx++) {
+
+            EGS_Vector transformed(point);
+            transforms[sidx]->transform(transformed);
+            int base_reg = base->isWhere(transformed);
+            bool in_base = base_reg >= 0;
+
+            if (in_base) {
+                results.regions_with_inscribed[base_reg].insert(inscribed[sidx]);
+                hit_counter[base_reg] += 1;
+            }
+
+        }
+    }
+
+    results.inscribed_volume = opts->bounds_volume*(double)n_in_inscribed/(int)opts->npoints;
+
+    if (opts->mode != NO_CORRECTION) {
+        results.corrected_volumes = applyVolumeCorrections(opts, hit_counter, results.uncorrected_volumes);
+    }
+    else {
+        results.corrected_volumes = results.uncorrected_volumes;
+    }
+
+    results.success = true;
+
+    results.time = timer.time();
+
+    return results;
+
+}
+
+
+bool isGZip(istream &vfile) {
+    return (vfile.get() == 0x1f && vfile.get() == 0x8b);
+}
+
+
+void readVolumes(istream &vfile, vector<RegVolume> &reg_volumes, RegionGeomSetT &reg_with_inscribed, vector<EGS_BaseGeometry *> inscribed) {
+    int nrecords;
+    vfile >> nrecords;
+    for (int rec = 0; rec < nrecords; rec++) {
+        int reg, ninscribed;
+        EGS_Float vol;
+        vfile >> reg >> vol >> ninscribed;
+        egsInformation("loaded %d/%d %f %d\n", reg, nrecords, vol, ninscribed);
+        reg_volumes.push_back(RegVolume(reg, vol));
+        for (int i=0; i < ninscribed; i++) {
+            int gidx;
+            vfile >> gidx;
+            reg_with_inscribed[reg].insert(inscribed[gidx]);
+        }
+    }
+
+}
+
+int loadVolumes(string fname, vector<RegVolume> &reg_volumes, RegionGeomSetT &reg_with_inscribed, vector<EGS_BaseGeometry *> inscribed) {
+    ifstream vfile(fname.c_str(), ios::binary);
```

Comment:
This fails for "text" format volume files, because they're not binary. Try something like:

```
int loadVolumes(string fname, vector<RegVolume> &reg_volumes, RegionGeomSetT &reg_with_inscribed, vector<EGS_BaseGeometry *> inscribed) {
    ifstream vfile(fname.c_str(), ios::binary);
    if (!vfile.is_open()) {
        return 1;
    }
    
    if (isGZip(vfile)) {
        vfile.close();
#ifdef HAS_GZSTREAM
        igzstream gzf(fname.c_str());
        readVolumes(gzf, reg_volumes, reg_with_inscribed, inscribed);
        gzf.close();
#else
        egsWarning("Tried to load gzip volume correction file but not compiled with gzstream.\n");
        return 1;
#endif
    }
    else {
        vfile.close();
        ifstream vfile2(fname.c_str(), ios::in);
        readVolumes(vfile2, reg_volumes, reg_with_inscribed, inscribed);
        vfile2.close();
    }

    return 0;

}
```